File: film_tuning.py
# SPDX-License-Identifier: GPL-3.0-only
"""Versioned LightTable interpretations; separate from measured film profiles.

V1 gently contracts the source yellow-green hue interval toward green before
spectral reconstruction. True yellows, skin, neutrals, blues and cyans are
outside its support. Luminance is preserved in linear ProPhoto (D50).
This is a visual interpretation with numerical guardrails, not a film scan fit.
"""
from __future__ import annotations

import hashlib
import json

import numpy as np

TUNINGS = {
    "kodak_portra_160": {"green_amount": 0.9},
    "kodak_portra_400": {"green_amount": 0.8},
    "kodak_portra_800": {"green_amount": 0.55},
}
VERSION = "1"
DESCRIPTION = ("LightTable's interpretation, adjusted using reference "
               "photographs and color checks. Not calibrated to a measured film scan.")

# Processed sources (JPEG, HEIC, TIFF) arrive display-referred: a camera or an
# editor has already applied a tone curve, so their mid-tones are lifted and
# their highlights are packed into the top of the range. The film model expects
# scene light and adds its own toe and shoulder, which is why a finished JPEG
# rendered as-is comes out flat and grey. Before filming, the expansion below
# raises the linear values to this power around middle grey: grey stays put,
# shadows fall back to where a scene would have them, and highlights extend
# above 1.0 as scene highlights do. Metering is unaffected at middle grey, so
# an edit with a fixed print exposure keeps its brightness. RAW decodes are
# scene-linear already and are never expanded.
DISPLAY_EXPANSION = 1.8
MIDDLE_GREY_LINEAR = 0.18

# Bradford-adapted linear ProPhoto D50 -> linear sRGB D65, as used by
# colour-science. Only the inverse's red column is needed for this hue interval.
TO_SRGB = np.asarray([
    [2.0340626798588866, -0.7275927647469631, -0.30682606926988176],
    [-0.22868941146731234, 1.2317680014826924, -0.0028926564624774157],
    [-0.008499732546934672, -0.15328915068335006, 1.1615688759268012],
], dtype=np.float64)
FROM_SRGB_RED = np.asarray([
    0.5293363376743782, 0.09831587646200844, 0.016847881261919936,
], dtype=np.float64)
PROPHOTO_Y = np.asarray([0.2880402, 0.7118741, 0.0000857])

TUNING_DIGEST = hashlib.sha256(json.dumps({
    "version": VERSION, "stocks": TUNINGS, "matrix": TO_SRGB.tolist(),
    "red": FROM_SRGB_RED.tolist(), "luminance": PROPHOTO_Y.tolist(),
    "algorithm": "source-yellow-green-contraction-v1",
    "display_expansion": DISPLAY_EXPANSION, "middle_grey": MIDDLE_GREY_LINEAR,
}, sort_keys=True).encode()).hexdigest()


def profile_tunings(stock: str) -> list[dict]:
    if stock not in TUNINGS:
        return []
    return [{"id": "lighttable", "version": VERSION,
             "name": "LightTable tuned", "description": DESCRIPTION}]


def specification(params: dict) -> dict | None:
    """The input preparation both engines apply before filming, or None.

    Two independent parts share one specification so the pixels are prepared
    exactly once: the versioned green interpretation, which only the tuned
    stocks carry, and the display-referred expansion, which every processed
    source needs and no RAW decode does. A RAW source with an untuned stock
    is the one case that needs nothing, and returns None so the engines take
    their original path byte for byte.
    """
    if params.get("profile_enabled", True) is False:
        return None
    tuned = (params.get("film_tuning", "original") == "lighttable"
             and str(params.get("film_tuning_version", VERSION)) == VERSION
             and params.get("stock") in TUNINGS)
    linear_input = bool(params.get("linear_input", False))
    if not tuned and linear_input:
        return None
    return {"version": 1,
            "green_amount": TUNINGS[params["stock"]]["green_amount"] if tuned else 0.0,
            "input_cctf_decoding": not linear_input,
            "display_expansion": 0.0 if linear_input else DISPLAY_EXPANSION}


def _smooth(value):
    value = np.clip(value, 0.0, 1.0)
    return value * value * (3.0 - 2.0 * value)


def prepare_input(image: np.ndarray, spec: dict) -> np.ndarray:
    """Return linear ProPhoto input without mutating the source/cache array.

    Bounded chunks avoid several full-resolution float64 intermediates. Rust's
    resident renderer implements the same operation before all film stages.
    Encoded processed sources use the ROMM/ProPhoto transfer curve, not sRGB.
    """
    if spec.get("version") != 1:
        raise ValueError("Unsupported LightTable film tuning version")
    amount = float(spec["green_amount"])
    if not np.isfinite(amount) or not 0 <= amount < 1:
        raise ValueError("Invalid LightTable green amount")
    expansion = float(spec.get("display_expansion", 0.0))
    if not np.isfinite(expansion) or expansion < 0 or expansion > 4:
        raise ValueError("Invalid LightTable display expansion")
    source = np.asarray(image, dtype=np.float32)
    if source.ndim != 3 or source.shape[-1] != 3:
        raise ValueError("Film tuning requires RGB input")
    out = np.empty_like(source)
    source_flat, out_flat = source.reshape(-1, 3), out.reshape(-1, 3)
    for start in range(0, len(source_flat), 262144):
        stop = min(start + 262144, len(source_flat))
        linear = source_flat[start:stop].astype(np.float64)
        if not np.isfinite(linear).all():
            raise ValueError("Film tuning requires finite input")
        if spec.get("input_cctf_decoding", False):
            linear = np.where(linear < 0.03125, linear / 16.0,
                              np.maximum(linear, 0.0) ** 1.8)
        if expansion > 0:
            # Anchored at middle grey, so exposure is unchanged and highlights
            # are free to exceed 1.0. Non-positive values are left alone.
            linear = np.where(
                linear > 0,
                MIDDLE_GREY_LINEAR * (np.maximum(linear, 0.0) / MIDDLE_GREY_LINEAR) ** expansion,
                linear)
        rgb = linear @ TO_SRGB.T
        r, g, b = rgb.T
        delta = g - b
        # This sector has G=max, B=min. Other sectors are exact no-ops.
        eligible = (g > r) & (r > b) & (b >= 0) & (g > 1e-10)
        hue = 120.0 + 60.0 * np.divide(b - r, delta,
            out=np.zeros_like(delta), where=delta > 1e-10)
        saturation = np.divide(delta, g, out=np.zeros_like(g), where=g > 1e-10)
        weight = _smooth((hue - 60.0) / 30.0) * _smooth(saturation / 0.15)
        shift = np.where(eligible,
            delta * (120.0 - hue) / 60.0 * amount * weight, 0.0)
        candidate = linear - shift[:, None] * FROM_SRGB_RED
        before, after = linear @ PROPHOTO_Y, candidate @ PROPHOTO_Y
        gain = np.divide(before, after, out=np.ones_like(before), where=after > 1e-10)
        candidate *= gain[:, None]
        # Preserve luminance while easing back toward the original at the
        # input gamut boundary. No per-channel clipping or hue discontinuity.
        difference = candidate - linear
        room = np.divide(1.0 - linear, difference,
            out=np.ones_like(linear), where=difference > 1e-10)
        blend = np.clip(np.min(room, axis=1), 0.0, 1.0)
        tuned = linear + blend[:, None] * difference
        out_flat[start:stop] = np.where((shift > 0)[:, None], tuned, linear)
    return out

File: rust-engine/src/film_tuning.rs
// SPDX-License-Identifier: GPL-3.0-only
//! Versioned LightTable interpretation of the source before spectral filming.
//! Keep this operation in sync with `film_tuning.prepare_input`. It works on
//! linear ProPhoto input, preserves its luminance, and leaves colors outside
//! the linear-sRGB yellow-green sector untouched.

use anyhow::{Result, bail};
use rayon::prelude::*;
use serde::{Deserialize, Serialize};
use spektrafilm_math::{image::ImageBuf, precision};

/// Middle grey in linear light; the fixed point of the display expansion.
const MIDDLE_GREY_LINEAR: f64 = 0.18;

#[derive(Debug, Clone, Copy, Deserialize, Serialize, PartialEq)]
#[serde(deny_unknown_fields)]
pub struct Specification {
    pub version: u32,
    pub green_amount: f64,
    pub input_cctf_decoding: bool,
    /// Power applied to processed (display-referred) sources around middle
    /// grey before filming, so a finished JPEG's compressed highlights and
    /// lifted shadows go back to scene proportions. Zero leaves the input
    /// alone; RAW decodes always send zero.
    #[serde(default)]
    pub display_expansion: f64,
    /// Linear luminance the expansion pivots on. The application measures it
    /// so the film meter's centre-weighted mean is unchanged by the expansion;
    /// middle grey is the fallback.
    #[serde(default = "default_anchor")]
    pub display_expansion_anchor: f64,
}

fn default_anchor() -> f64 {
    MIDDLE_GREY_LINEAR
}

impl Specification {
    pub fn validate(&self) -> Result<()> {
        if self.version != 1 {
            bail!("unsupported LightTable film tuning version");
        }
        if !self.green_amount.is_finite() || !(0.0..1.0).contains(&self.green_amount) {
            bail!("invalid LightTable green amount");
        }
        if !self.display_expansion.is_finite() || !(0.0..=4.0).contains(&self.display_expansion) {
            bail!("invalid LightTable display expansion");
        }
        if !self.display_expansion_anchor.is_finite()
            || !(0.001..=4.0).contains(&self.display_expansion_anchor)
        {
            bail!("invalid LightTable display expansion anchor");
        }
        Ok(())
    }

    /// Absent tuning retains every existing cache key byte for byte.
    pub fn extend_key(&self, original: String) -> String {
        format!(
            "{original}:input_tuning:{}",
            serde_json::to_string(self).expect("validated finite tuning specification")
        )
    }
}

const TO_SRGB: [[f64; 3]; 3] = [
    [
        2.0340626798588866,
        -0.7275927647469631,
        -0.30682606926988176,
    ],
    [
        -0.22868941146731234,
        1.2317680014826924,
        -0.0028926564624774157,
    ],
    [
        -0.008499732546934672,
        -0.15328915068335006,
        1.1615688759268012,
    ],
];
const FROM_SRGB_RED: [f64; 3] = [
    0.5293363376743782,
    0.09831587646200844,
    0.016847881261919936,
];
const PROPHOTO_Y: [f64; 3] = [0.2880402, 0.7118741, 0.0000857];

fn dot(a: [f64; 3], b: [f64; 3]) -> f64 {
    a[0] * b[0] + a[1] * b[1] + a[2] * b[2]
}

fn smooth(value: f64) -> f64 {
    let value = value.clamp(0.0, 1.0);
    value * value * (3.0 - 2.0 * value)
}

fn prepare_pixel(mut linear: [f64; 3], spec: Specification) -> [f64; 3] {
    if spec.input_cctf_decoding {
        for channel in &mut linear {
            // ROMM/ProPhoto transfer function, including its linear toe.
            *channel = if *channel < 0.03125 {
                *channel / 16.0
            } else {
                channel.max(0.0).powf(1.8)
            };
        }
    }
    if spec.display_expansion > 0.0 {
        for channel in &mut linear {
            // Anchored at middle grey, so exposure is unchanged and highlights
            // are free to exceed 1.0. Non-positive values are left alone.
            if *channel > 0.0 {
                let anchor = spec.display_expansion_anchor;
                *channel = anchor * (channel.max(0.0) / anchor).powf(spec.display_expansion);
            }
        }
    }
    let [r, g, b] = TO_SRGB.map(|row| dot(row, linear));
    if !(g > r && r > b && b >= 0.0 && g > 1e-10) {
        return linear;
    }
    let delta = g - b;
    let hue = 120.0 + 60.0 * if delta > 1e-10 { (b - r) / delta } else { 0.0 };
    let saturation = delta / g;
    let weight = smooth((hue - 60.0) / 30.0) * smooth(saturation / 0.15);
    let shift = delta * (120.0 - hue) / 60.0 * spec.green_amount * weight;
    if shift <= 0.0 {
        return linear;
    }
    let mut candidate = std::array::from_fn(|c| linear[c] - shift * FROM_SRGB_RED[c]);
    let before = dot(linear, PROPHOTO_Y);
    let after = dot(candidate, PROPHOTO_Y);
    let gain = if after > 1e-10 { before / after } else { 1.0 };
    for channel in &mut candidate {
        *channel *= gain;
    }
    let difference: [f64; 3] = std::array::from_fn(|c| candidate[c] - linear[c]);
    let blend = (0..3)
        .map(|c| {
            if difference[c] > 1e-10 {
                (1.0 - linear[c]) / difference[c]
            } else {
                1.0
            }
        })
        .fold(1.0_f64, f64::min)
        .clamp(0.0, 1.0);
    std::array::from_fn(|c| linear[c] + blend * difference[c])
}

pub fn prepare_input(source: &ImageBuf, spec: Specification) -> Result<ImageBuf> {
    spec.validate()?;
    if !source
        .data
        .par_iter()
        .all(|&value| precision::to_f64(value).is_finite())
    {
        bail!("film tuning requires finite input");
    }
    let mut output = ImageBuf::new(source.width, source.height);
    output
        .data
        .par_chunks_exact_mut(3)
        .zip(source.data.par_chunks_exact(3))
        .for_each(|(out, pixel)| {
            let result = prepare_pixel(
                std::array::from_fn(|c| {
                    // Python's input is explicitly float32, even in precision-f64 builds.
                    f64::from(precision::to_f32(pixel[c]))
                }),
                spec,
            );
            for c in 0..3 {
                out[c] = precision::from_f32(result[c] as f32);
            }
        });
    Ok(output)
}

#[cfg(test)]
mod tests {
    use super::*;

    fn spec() -> Specification {
        Specification {
            version: 1,
            green_amount: 0.9,
            input_cctf_decoding: false,
            display_expansion: 0.0,
            display_expansion_anchor: MIDDLE_GREY_LINEAR,
        }
    }

    #[test]
    fn display_expansion_holds_grey_and_frees_highlights() {
        // Values captured from film_tuning.prepare_input with expansion 1.8,
        // green amount 0, on float32 sources.
        let expanded = Specification {
            green_amount: 0.0,
            display_expansion: 1.8,
            ..spec()
        };
        let cases: [([f32; 3], [f64; 3]); 4] = [
            ([0.18, 0.18, 0.18], [0.18000000715255737; 3]),
            ([1.0, 0.5, 0.05], [3.9425933361053467, 1.1322126388549805, 0.01794436201453209]),
            ([0.0, -0.01, 0.02], [0.0, -0.009999999776482582, 0.003448545467108488]),
            ([0.36, 0.09, 0.72], [0.6267964243888855, 0.05169142782688141, 2.1826319694519043]),
        ];
        for (pixel, expected) in cases {
            // prepare_input feeds prepare_pixel float32-rounded values.
            let actual = prepare_pixel(pixel.map(f64::from), expanded);
            for c in 0..3 {
                assert!((actual[c] - expected[c]).abs() < 1e-6, "{pixel:?} -> {actual:?}");
            }
        }
        assert!(Specification { display_expansion: 4.5, ..spec() }.validate().is_err());
        assert!(Specification { display_expansion: -0.1, ..spec() }.validate().is_err());
        let absent: Specification = serde_json::from_str(
            r#"{"version":1,"green_amount":0.8,"input_cctf_decoding":true}"#,
        )
        .unwrap();
        assert_eq!(absent.display_expansion, 0.0);
        assert_eq!(absent.display_expansion_anchor, MIDDLE_GREY_LINEAR);
        // The anchor is the fixed point of the expansion whatever its value.
        let anchored = Specification { green_amount: 0.0, display_expansion: 1.8, display_expansion_anchor: 0.42, ..spec() };
        let held = prepare_pixel([0.42, 0.42, 0.42], anchored);
        assert!(held.iter().all(|v| (v - 0.42).abs() < 1e-12));
        assert!(Specification { display_expansion_anchor: 0.0, ..spec() }.validate().is_err());
    }

    #[test]
    fn python_reference_pixels_match_for_linear_and_romm_inputs() {
        // Values captured from film_tuning.prepare_input on float32 sources.
        let pixels = [
            [0.22, 0.32, 0.07],
            [0.98, 0.99, 0.55],
            [0.18, 0.18, 0.18],
            [0.45, 0.2, 0.08],
            [0.05, 0.2, 0.35],
            [-0.01, 0.1, 0.05],
            [0.03125, 0.02, 0.01],
        ];
        let expected_linear = [
            [0.16070108115673065, 0.3439929485321045, 0.07601317018270493],
            [0.9800000190734863, 0.9900000095367432, 0.550000011920929],
            [0.18000000715255737; 3],
            [
                0.44999998807907104,
                0.20000000298023224,
                0.07999999821186066,
            ],
            [0.05000000074505806, 0.20000000298023224, 0.3499999940395355],
            [
                -0.009999999776482582,
                0.10000000149011612,
                0.05000000074505806,
            ],
            [0.03125, 0.019999999552965164, 0.009999999776482582],
        ];
        let expected_encoded = [
            [
                0.06551802903413773,
                0.1286085844039917,
                0.008340200409293175,
            ],
            [0.9642884135246277, 0.9820720553398132, 0.3409202992916107],
            [0.045655231922864914; 3],
            [0.2375650256872177, 0.055189188569784164, 0.0106062525883317],
            [
                0.004551410675048828,
                0.055189188569784164,
                0.15112018585205078,
            ],
            [
                -0.0006249999860301614,
                0.015848932787775993,
                0.004551410675048828,
            ],
            [0.001953125, 0.0012499999720603228, 0.0006249999860301614],
        ];
        let source = ImageBuf::from_data(
            7,
            1,
            pixels
                .into_iter()
                .flatten()
                .map(precision::from_f32)
                .collect(),
        );
        for (decode, expected) in [(false, expected_linear), (true, expected_encoded)] {
            let mut specification = spec();
            specification.input_cctf_decoding = decode;
            let actual = prepare_input(&source, specification).unwrap();
            for (&actual, expected) in actual.data.iter().zip(expected.into_iter().flatten()) {
                assert!((precision::to_f64(actual) - expected).abs() < 1e-7);
            }
        }
        assert_eq!(precision::to_f32(source.data[0]), 0.22);
    }

    #[test]
    fn green_adjustment_preserves_luminance_and_stays_in_gamut() {
        let source = [0.22, 0.32, 0.07];
        for amount in [0.0, 0.55, 0.8, 0.9, 0.999] {
            for exposure in [0.1, 1.0, 2.0, 3.0] {
                let pixel = source.map(|value| value * exposure);
                let tuned = prepare_pixel(
                    pixel,
                    Specification {
                        green_amount: amount,
                        ..spec()
                    },
                );
                assert!((dot(pixel, PROPHOTO_Y) - dot(tuned, PROPHOTO_Y)).abs() < 1e-12);
                assert!(tuned.iter().all(|&value| (0.0..=1.0).contains(&value)));
                assert!(tuned[0] <= pixel[0]);
                assert!(tuned[1] >= pixel[1]);
            }
        }
    }

    #[test]
    fn zero_amount_and_outside_sector_are_exact_noops() {
        for pixel in [
            [0.18; 3],
            [0.45, 0.2, 0.08],
            [0.05, 0.2, 0.35],
            [0.5, 0.5, 0.03],
            [-0.01, 0.1, 0.05],
        ] {
            assert_eq!(prepare_pixel(pixel, spec()), pixel);
        }
        let pixel = [0.22, 0.32, 0.07];
        assert_eq!(
            prepare_pixel(
                pixel,
                Specification {
                    green_amount: 0.0,
                    ..spec()
                }
            ),
            pixel
        );
    }

    #[test]
    fn invalid_specifications_and_pixels_are_rejected() {
        for amount in [-0.1, 1.0, f64::NAN, f64::INFINITY] {
            assert!(
                Specification {
                    green_amount: amount,
                    ..spec()
                }
                .validate()
                .is_err()
            );
        }
        assert!(
            Specification {
                version: 2,
                ..spec()
            }
            .validate()
            .is_err()
        );
        let source = ImageBuf::from_data(1, 1, vec![precision::from_f32(f32::NAN); 3]);
        assert!(prepare_input(&source, spec()).is_err());
    }
}
